etcd/watchmanager/manager.go

```diff
@@ -139,6 +145,15 @@ func (w *manager) Watch(key string) {
 				w.m.etcdWatchError.Inc(1)
 				// do not stop here, even though the update contains an error
 				// we still take this chance to attempt a Get() for the latest value
+
+				// If the current revision has been compacted, set watchChan to
+				// nil so the watch is recreated with a valid start revision
+				if err == rpctypes.ErrCompacted {
+					w.logger.Warnf("recreating watch at revision %d", r.CompactRevision)
+					startRev = r.CompactRevision
+					watchChan = nil
+					continue
```

Comment:
Is this guaranteed to get the watch created at some point? By `continue` here, you basically skipped the `updateFn` logic which would update the kv.Watch with latest value or local cached value when the watch could not be watched, we should probably keep that behavior in case the watcher creation never success. I guess you can get it by removing the continue? Please add a unit test for this case as well.